Initial contribution of the LeapMotion binding

kaikreuzer opened the pull request:
This has been originally written by @teichsta in 2013 in a Boing 747 on the flight to JavaOne.
The [demo video](https://www.youtube.com/watch?v=czq3S13tGVk) was produced in our hotel corridor in SF.

It was never finished and thus not contributed to openHAB 1. I now finally took the time to finish the code, adapt it to openHAB 2 and made use of the very latest concepts such as trigger channels and configurable profiles.

Also-by: Thomas Eichstädt-Engelen <thomas@openhab.org>
Signed-off-by: Kai Kreuzer <kai@openhab.org>

wborn commented on addons/binding/org.openhab.binding.leapmotion/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,34 @@
+Service-Component: OSGI-INF/*.xml
+Bundle-NativeCode: 
+ lib/libLeapJava.dylib;osname=macosx; processor=x86_64,
+ lib/libLeap.dylib;osname=macosx; processor=x86_64
```
Will the build servers be able to build this? It fails on my x86_64 with Ubuntu:

```
[ERROR] Internal error: java.lang.RuntimeException: org.osgi.framework.BundleException: Bundle org.openhab.binding.leapmotion cannot be resolved
[ERROR] Resolution errors:
[ERROR]    Bundle org.openhab.binding.leapmotion - No match found for native code: lib/libLeapJava.dylib; processor=x86_64; osname=macosx, lib/libLeap.dylib; processor=x86_64; osname=macosx
[ERROR] -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: org.osgi.framework.BundleException: Bundle org.openhab.binding.leapmotion cannot be resolved
Resolution errors:
   Bundle org.openhab.binding.leapmotion - No match found for native code: lib/libLeapJava.dylib; processor=x86_64; osname=macosx, lib/libLeap.dylib; processor=x86_64; osname=macosx

    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:120)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: java.lang.RuntimeException: org.osgi.framework.BundleException: Bundle org.openhab.binding.leapmotion cannot be resolved
Resolution errors:
   Bundle org.openhab.binding.leapmotion - No match found for native code: lib/libLeapJava.dylib; processor=x86_64; osname=macosx, lib/libLeap.dylib; processor=x86_64; osname=macosx

    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.getResolverState (OsgiBundleProject.java:265)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.resolveClassPath (OsgiBundleProject.java:164)
    at org.eclipse.tycho.core.resolver.DefaultTychoResolver.resolveProject (DefaultTychoResolver.java:130)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.afterProjectsRead (TychoMavenLifecycleParticipant.java:95)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:264)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.osgi.framework.BundleException: Bundle org.openhab.binding.leapmotion cannot be resolved
Resolution errors:
   Bundle org.openhab.binding.leapmotion - No match found for native code: lib/libLeapJava.dylib; processor=x86_64; osname=macosx, lib/libLeap.dylib; processor=x86_64; osname=macosx

    at org.eclipse.tycho.core.osgitools.EquinoxResolver.assertResolved (EquinoxResolver.java:305)
    at org.eclipse.tycho.core.osgitools.EquinoxResolver.newResolvedState (EquinoxResolver.java:75)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.getResolverState (OsgiBundleProject.java:262)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.resolveClassPath (OsgiBundleProject.java:164)
    at org.eclipse.tycho.core.resolver.DefaultTychoResolver.resolveProject (DefaultTychoResolver.java:130)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.afterProjectsRead (TychoMavenLifecycleParticipant.java:95)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:264)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/InternalErrorException
```

wborn commented on addons/binding/org.openhab.binding.leapmotion/src/main/java/org/openhab/binding/leapmotion/LeapMotionBindingConstants.java:
```diff
@@ -0,0 +1,44 @@
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
```
Actually I was trying to build the project to see the SAT results. That would have probably shown that we still use EPL-1.0 instead of EPL-2.0 in this repo. ;-) Maybe you can check the SAT results for more issues?

wborn commented on addons/binding/org.openhab.binding.leapmotion/pom.xml:
```diff
@@ -0,0 +1,18 @@
+	<parent>
+		<groupId>org.openhab.binding</groupId>
+		<artifactId>pom</artifactId>
+		<version>2.4.0-SNAPSHOT</version>
```
Can you also add a module to the parent POM and add a feature for it?

kaikreuzer commented:
Haha, I did all the stupid mistakes I always complain about in reviews - could to be on the other side of the fence this time :-)
Will include the binding where it is required and check SAT and the build...

kaikreuzer commented:
Thanks for the very quick review, @wborn!
I hope to have addressed all your comments. The build issue should be fixed as well now, let's see what Travis says.

wborn commented:
Yes the code builds again on Linux. Thanks! I also see there's not much for SAT to report on. :-)

wborn commented on addons/binding/org.openhab.binding.leapmotion/src/main/java/org/openhab/binding/leapmotion/handler/LeapMotionHandler.java:
```diff
@@ -0,0 +1,166 @@
+ * @author Thomas Eichstädt-Engelen - Initial version of listener logic
+ *
+ */
+public class LeapMotionHandler extends BaseThingHandler {
```
This is the only class that did not get the `@NonNullByDefault` treatment against NPEs. ;-)

kaikreuzer replied:
Yeah, the skeleton didn't create that at the time that I used it.
I've adapted the code now.

wborn commented on addons/binding/org.openhab.binding.leapmotion/README.md:
```diff
@@ -0,0 +1,54 @@
+# LeapMotion Binding
+
+The LeapMotion controller is a gesture sensoring device that uses stereoscopic cameras and is connected through USB.
```
* Can you also add a link to the [Leap Motion](https://www.leapmotion.com/) website here so it's easy to find more information about the company/device?
* On the website they refer to themselves as "Leap Motion" and not "LeapMotion" so you might want to update all references for consistency.
* Please place each sentence on a new line in this Markdown file (where possible) so it's easier to add review comments.
* Can you also fix the remaining (minor) SAT issues? :-)

I think this is the first binding that uses profiles so it makes a good example! :+1:

kaikreuzer commented:
@wborn Anything still open here? Note that the Jenkins failure is not due to this PR and Travis is happy with it :-)

kaikreuzer commented:
Excellent, then don't wait any longer 😎 .

wborn commented:
Many thanks for the binding!
